resharper/resharper-unity/src/Yaml/Psi/Modules/UnityYamlDisableStrategy.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Collections.Generic;
+using JetBrains.Application.Settings;
+using JetBrains.DataFlow;
+using JetBrains.ProjectModel;
+using JetBrains.ProjectModel.DataContext;
+using JetBrains.ReSharper.Plugins.Unity.Settings;
+using JetBrains.Util;
+
+namespace JetBrains.ReSharper.Plugins.Unity.Yaml.Psi.Modules
+{
+    [SolutionComponent]
+    public class UnityYamlDisableStrategy
+    {
+        private long myTotalSize = 0;
+
+        private readonly IProperty<bool> myEnabled;
+        private readonly AssetSerializationMode myAssetSerializationMode;
+        protected readonly IProperty<bool> YamlParsingEnabled;
+        private const long YamlFileSizeThreshold = 40 * (1024 * 1024); // 40 MB
+        private const long YamlFileTotalSizeThreshold = 700 * (1024 * 1024); // 700 MB
+
+        public UnityYamlDisableStrategy(Lifetime lifetime, ISolution solution, ISettingsStore settingsStore, 
+            AssetSerializationMode assetSerializationMode, UnityYamlEnabled unityYamlEnabled)
+        {
+            myAssetSerializationMode = assetSerializationMode;
+            var boundStore = settingsStore.BindToContextLive(lifetime, ContextRange.ManuallyRestrictWritesToOneContext(solution.ToDataContext()));
+            myEnabled = boundStore.GetValueProperty(lifetime, (UnitySettings s) => s.EnableYamlHeuristic);
+            YamlParsingEnabled = unityYamlEnabled.YamlParsingEnabled;
+
+        }
+
+        private bool IsYamlParsingAvailable()
+        {
+            return myAssetSerializationMode.IsForceText && YamlParsingEnabled.Value;
+        }
+        
+        public void Run(FileSystemPath solutionDirectory)
+        {
+            if (myEnabled.Value)
+            {
+                if (!IsYamlParsingAvailable())
+                {
+                    YamlParsingEnabled.Value = false;
+                } 
+                else if (IsAnyFilePreventYamlParsing(solutionDirectory) || myTotalSize > YamlFileTotalSizeThreshold)
+                {
+                    YamlParsingEnabled.Value = false;
+                    CreateNotification();
+                }
+                myEnabled.Value = false;
```

Comment:
Does this mean we only ever run the heuristic once per solution? What if a file is added that is e.g. too large? If it is intended to be a run once setting, the name in the setting file should reflect that - `EnableYamlHeuristic` doesn't indicate that it's only run once.

Reply:
Yes.
We discussed it today with @korifey and decided that if user adds very large file, he understands that Rider processing this file. 
Also, it will be very strange when yaml features are enabled and after some time (for user, he doesn't know our idea with disabling) yaml features are disabled

Reply:
I don't think that's the right approach - if someone else in the team adds a large scene, or in 6 months time a scene has grown in complexity, we'll ignore it. I think if we're going to have this heuristic, we should run it every time. I think it can easily be a part of the initial call to `OnUnityProjectAdded`